src/base.jl

```diff
@@ -1 +1,34 @@
+
 @functor Base.RefValue
+
+functor(::Type{<:Base.ComposedFunction}, x) = (outer = x.outer, inner = x.inner), y -> Base.ComposedFunction(y.outer, y.inner)
+
+using LinearAlgebra
+# The reason for these is to let W and W' be seen as tied weights in Flux models.
+# Can't treat ReshapedArray very well, as its type doesn't include enough details for reconstruction.
+
+functor(::Type{<:Adjoint}, x) = (parent = _adjoint(x),), y -> adjoint(only(y))
+
+_adjoint(x) = adjoint(x)  # _adjoint is the inverse, and also understands more types:
+_adjoint(x::NamedTuple{(:parent,)}) = x.parent  # "structural" gradient, and lazy broadcast used by Optimisers:
+_adjoint(bc::Broadcast.Broadcasted{S}) where S = Broadcast.Broadcasted{S}(_conjugate(bc.f, adjoint), _adjoint.(bc.args))
+
+functor(::Type{<:Transpose}, x) = (parent = _transpose(x),), y -> transpose(only(y))
+
+_transpose(x) = transpose(x)
+_transpose(x::NamedTuple{(:parent,)}) = x.parent
+_transpose(bc::Broadcast.Broadcasted{S}) where S = Broadcast.Broadcasted{S}(_conjugate(bc.f, transpose), _transpose.(bc.args))
```

Comment:
Some chance this Broadcasted stuff is too Optimiser-specific. But piracy to put it there? The less elaborate version is to give up and `materialize(bc)` like PermutedDimsArray here.

Ideally, I think this would actually move to ChainRulesCore, as some kind of BroadcastThunk object. But... maybe later.

Reply:
Why would it be piracy to include in Optimisers over Functors? One way to avoid it could be to pull the thread a little harder and bring anything connected to it over to Optimisers as well. Depends how much we see Functors as a minimal "core" library vs something that should provide a reasonable OOTB experience.

Reply:
I literally mean that `Functors._transpose(bc::Broadcast....)` is piracy.